certbot/certbot/_internal/client.py

```diff
@@ -449,13 +456,17 @@ def obtain_certificate(self, domains: List[str], old_keypath: Optional[str] = No
         # domains contains a wildcard because the ACME spec forbids identifiers
         # in authzs from containing a wildcard character.
         if self.config.allow_subset_of_names and successful_domains != domains:
-            if not self.config.dry_run:
-                os.remove(key.file)
-                os.remove(csr.file)
-            return self.obtain_certificate(successful_domains)
+            return self._retry_obtain_certificate(key, csr, domains, successful_domains)
         else:
-            cert, chain = self.obtain_certificate_from_csr(csr, orderr)
-            return cert, chain, key, csr
+            try:
+                cert, chain = self.obtain_certificate_from_csr(csr, orderr)
+                return cert, chain, key, csr
+            except messages.Error as error:
```

Comment:
It would be good to have a comment here to explain what is happening. i.e. Some domains may be rejected during the very late stage of order finalization and Certbot can retry the operation without the rejected domains.

Reply:
Done!